readme.md
yaml for lua [![Build Status](https://travis-ci.org/lubyk/yaml.png)](https://travis-ci.org/lubyk/yaml)
===========

Very fast yaml parser based on libYAML 0.1.3 by Kirill Simonov

[Documentation](http://doc.lubyk.org/yaml.html).

install
-------

    luarocks install yaml


History.txt
== 1.1.2 2015-05-11

  * Support for lua 5.3.

== 1.1.1 2014-08-15

  * Fixed compilation on linux.

== 1.1.0 2014-08-15

  * Added support for sorted table keys on dump.
  * Fixed lub dependency information.
  * Using lut.Builder to generate build files.

== 1.0.0 2014-05-08

  * Initial release based on libYAML 0.1.3, Andrew's bindings '0.2'.
  * Lua 5.2 compatibility.
  * Removed external dependency.
  * Added 'yaml.loadpath', documentation and testing.
  * Added option to 'safe' parse (ignores references in yaml).
  * Fixed number parsing code to not rely on current locale.
  * Continuous integration testing with travis-ci.


test/yaml_test.lua
--[[------------------------------------------------------

  yaml test
  ---------

  ...

--]]------------------------------------------------------
local lub    = require 'lub'
local lut    = require 'lut'
local yaml   = require 'yaml'
local should = lut.Test('yaml')

function should.auto_load()
  assertTrue(yaml.load)
end

function should.load_hash()
  local simple = yaml.loadpath(lub.path '|fixtures/simple.yml')
  assertEqual('hello', simple.hash.a)
  assertEqual('lubyk', simple.hash.b)
end

function should.load_number()
  local simple = yaml.loadpath(lub.path '|fixtures/simple.yml')
  assertEqual(0.5, simple.number.a)
  assertEqual(3, simple.number.b)
end

function should.load_list()
  local simple = yaml.loadpath(lub.path '|fixtures/simple.yml')
  assertEqual('first',  simple.list[1])
  assertEqual('second', simple.list[2])
end

function should.loadpath()
  local simple = yaml.loadpath(lub.path '|fixtures/simple.yml')
  assertEqual('first',  simple.list[1])
  assertEqual('second', simple.list[2])
end

function should.useRefsAsSameObj()
  local refs = yaml.loadpath(lub.path '|fixtures/refs.yml')
  assertEqual('Jane',  refs.roles.boss.name)
  -- same obj
  assertTrue(refs.roles.wife == refs.roles.boss)
  assertTrue(refs.all[1]     == refs.roles.boss)
end

function should.parseMultipleValues()
  local a, b, c = yaml.load [[
--- 3
--- 4
--- 5
...
  ]]
  assertEqual(3, a)
  assertEqual(4, b)
  assertEqual(5, c)
end

function should.dump()
  assertMatch('b: 4', yaml.dump {a = {b = 4}})
end

function should.load()
  local data = yaml.load [[
hey: June
ok: true
  ]]
  assertEqual('June', data.hey)
  assertEqual(true, data.ok)
end

function should.dumpAndLoadAnchors()
  local l = { x = 1 }
  l.y = l
  local res = yaml.load(yaml.dump(l))
  assertEqual(res, res.y)
end

function should.disableAnchors()
  local l = { x = 1 }
  l.y = l
  -- true == safe loading
  local res = yaml.load(yaml.dump(l), true)
  assertNil(res.y)
end

function should.configure()
  assertPass(function()
    yaml.configure {
      load_set_metatables = false,
    }
  end)
end

function should.dumpSorted()
  yaml.configure {
    sort_table_keys = true,
  }
  local x = {
    dom     = 1,
    anna    = 2,
    louis   = 3,
    bernard = 4,
    armand  = 5,
    simon   = 6,
  }
  assertEqual([[
---
anna: 2
armand: 5
bernard: 4
dom: 1
louis: 3
simon: 6
]], yaml.dump(x))

  yaml.configure {
    sort_table_keys = false,
  }
end

function should.raiseErrorOnNonStringKeysAndSort()
  yaml.configure {
    sort_table_keys = true,
  }
  local x = {
    [{}]    = 0,
    dom     = 1,
    anna    = 2,
    louis   = 3,
    bernard = 4,
    armand  = 5,
    simon   = 6,
  }
  assertError('Key is not a string: cannot dump with sorted keys.', function()
    yaml.dump(x)
  end)

  yaml.configure {
    sort_table_keys = false,
  }
end
  
  

should:test()


scripts/rock.lua
#!/usr/bin/env lua
local lub = require 'lub'
local lib = require 'yaml'

local def = {
  description = {
    summary = "Very fast yaml parser based on libYAML by Kirill Simonov",
    author  = "Andrew Danforth, Gaspard Bucher",
    license = "MIT",

    detailed = [[
    This module is a Lua binding for Kirill Siminov's excellent LibYAML. LibYAML is generally considered to be the best C YAML 1.1 implementation.

    Main features are:
     - Fast and easy to use
     - Based on proven code (libYAML)
     - Support for table loops
     - No external dependencies

     Read the documentation at http://doc.lubyk.org/yaml.html.
    ]],
    homepage = "http://doc.lubyk.org/"..lib.type..".html",
  },

  includes  = {'src'},
  libraries = {},
  platlibs = {},
}

-- Platform specific sources or link libraries
def.platspec = def.platspec or lub.keys(def.platlibs)

--- End configuration

local tmp = lub.Template(lub.content(lub.path '|rockspec.in'))
lub.writeall(lib.type..'-'..lib.VERSION..'-1.rockspec', tmp:run {lib = lib, def = def, lub = lub})

tmp = lub.Template(lub.content(lub.path '|dist.info.in'))
lub.writeall('dist.info', tmp:run {lib = lib, def = def, lub = lub})

tmp = lub.Template(lub.content(lub.path '|CMakeLists.txt.in'))
lub.writeall('CMakeLists.txt', tmp:run {lib = lib, def = def, lub = lub})




yaml/init.lua
--[[------------------------------------------------------
  # YAML parser for Lua <a href="https://travis-ci.org/lubyk/yaml"><img src="https://travis-ci.org/lubyk/yaml.png" alt="Build Status"></a> 

  This parser uses [libYAML](http://pyyaml.org/wiki/LibYAML) by Kirill Siminov
  with bindings by Andrew Danforth and some modifications by Gaspard Bucher.
  libYAML version 0.1.3. LibYAML is generally considered to be the best C YAML 1.1 implementation. 

  <html><a href="https://github.com/lubyk/yaml"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_green_007200.png" alt="Fork me on GitHub"></a></html>

  *MIT license* &copy; Kirill Siminov 2006, Andrew Danforth 2009.

  ## Installation
  
  With [luarocks](http://luarocks.org):

    $ luarocks install yaml
  
  ## Usage example

    local data = yaml.load(some_yaml)

    local yaml_string = yaml.dump(some_table)

--]]-----------------------------------------------------
local lub     = require 'lub'
local core    = require 'yaml.core'
local lib     = lub.Autoload 'yaml'

local load, dump, configure = core.load, core.dump, core.configure

-- Current version respecting [semantic versioning](http://semver.org).
lib.VERSION = '1.1.2'

lib.DEPENDS = { -- doc
  -- Compatible with Lua 5.1 to 5.3 and LuaJIT
  'lua >= 5.1, < 5.4',
  -- Uses [Lubyk base library](http://doc.lubyk.org/lub.html)
  'lub >= 1.0.3, < 2',
}

-- nodoc
lib.DESCRIPTION = {
  summary = "Very fast yaml parser based on libYAML by Kirill Simonov",
  author  = "Andrew Danforth, Gaspard Bucher",
  license = "MIT",

  detailed = [[
  This module is a Lua binding for Kirill Siminov's excellent LibYAML. LibYAML is generally considered to be the best C YAML 1.1 implementation.

  Main features are:
   - Fast and easy to use
   - Based on proven code (libYAML)
   - Support for table loops
   - No external dependencies

   Read the documentation at http://doc.lubyk.org/yaml.html.
  ]],
  homepage = "http://doc.lubyk.org/yaml.html",
}

-- nodoc
lib.BUILD = {
  github = 'lubyk',
  includes  = {'src'},
  sources   = {'src/*.c'},
}

--[[

  # Lua table

  Lua tables works as sets. In Lua, there is no distinction between an array and
  a dictionary whereas YAML makes this distinction. When dumping a table, the
  library has to choose between these two types. Here is the selection rule:

  + array: The table contains an element in index position `1`.

  ## Lua

  Some lua code
  
    local data = {
      { job = 'Car wash',
        duration = 1.5,
        comment = 'The car was really filthy.'
      },
      { job = 'Create website',
        duration = 6.25,
        comment = 'Using bootstrap template.'
      }
    }
    -- Create a loop
    data[1].link = data

  ## YAML

  And the equivalent yaml using `yaml.dump(data)`:
  
    #yaml
    --- &0
    - link: *0
      duration: 1.5
      comment: The car was really filthy.
      job: Car wash
    - duration: 6.25
      comment: Using bootstrap template.
      job: Create website

--]]------------------------------------------------------

-- # Class methods

-- Parse a `string` containing yaml content and return lua values. If the second
-- argument `safe` is true, the library will not generate aliases in tables.
--
-- Note that if there are more then on value in the YAML content, this function
-- will return multiple values. Example:
--
--   local a, b, c = yaml.load [[
--   --- 3
--   --- 4
--   --- 5
--   ...
--   ]]
--   --> a = 3, b = 4, c = 5
--
-- function lib.load(string, safe)

-- nodoc
lib.load = load

-- Parse the YAML content of the file at `path` and return lua values. Uses
-- yaml.load internally.
function lib.loadpath(path)
  return load(lub.content(path))
end

-- Dump all lua values in the vararg as YAML. Note that when using 'load' on
-- the produced content, multiple values will be returned.
-- function lib.dump(...)

-- nodoc
lib.dump = dump

-- Configure parser. Pass a table setting options to true or false. The default
-- value is shown in parenthesis.
--
-- WARN Please note that this configuration alters parsing globaly and should be
-- avoided. If such configuration is often needed, please consult the maintainer
-- so that we can move such configuration inside a Parser object instead of
-- using global settings.
--
-- + dump_auto_array:           (true)
-- + dump_error_on_unsupported: (false)
-- + dump_check_metatables:     (true)
-- + load_set_metatables:       (true)
-- + load_numeric_scalars:      (true)
-- + load_nulls_as_nil:         (false)
-- + sort_table_keys:           (false)
-- function lib.configure(options)

-- nodoc
lib.configure = configure

return lib
